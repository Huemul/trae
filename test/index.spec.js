/* global describe it expect afterEach */

const fetchMock = require('fetch-mock');
const Trae      = require('../lib').Trae;

afterEach(() => {
  fetchMock.restore();
});

const TEST_URL = 'http://localhost:8080/foo';

describe('HTTP -> http', () => {
  it('Initilize default attributes on the constructor', () => {
    const trae = new Trae('/api');

    expect(trae._baseUrl).toEqual('/api');
    expect(trae._middleware).toBeDefined();
  });

  describe('get', () => {
    it('makes a GET request to baseURL + path', () => {
      const trae = new Trae();
      const url  = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status : 200,
        headers: {
          'Content-Type': 'application/json'
        },
        body: {
          foo: 'bar'
        }
      });

      return trae.get(url)
      .then((res) => {
        expect(res).toEqual({ foo: 'bar' });
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('get');
      });
    });
  });

  describe('del', () => {
    it('makes a DELETE request to baseURL + path', () => {
      const trae = new Trae();
      const url  = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status : 200,
        body   : 'Deleted!',
        headers: {
          'Content-Type': 'application/text'
        }
      }, {
        method: 'delete'
      });

      return trae.delete(url)
      .then((res) => {
        expect(res).toEqual('Deleted!');
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('delete');
      });
    });
  });

  describe('head', () => {
    it('makes a HEAD request to baseURL + path', () => {
      const trae = new Trae();
      const url  = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status: 200
      }, {
        method: 'head'
      });

      return trae.head(url)
      .then((res) => {
        expect(res).toEqual('');
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('head');
      });
    });
  });

  describe('post', () => {
    it('makes a POST request to baseURL + path', () => {
      const trae = new Trae();
      const url  = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status : 200,
        body   : { foo: 'bar' },
        headers: {
          'Content-Type': 'application/json'
        }
      }, {
        method: 'post'
      });

      return trae.post(url, { foo: 'bar' })
      .then((res) => {
        expect(res).toEqual({ foo: 'bar' });
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('post');
      });
    });
  });

  describe('put', () => {
    it('makes a PUT request to baseURL + path', () => {
      const trae = new Trae();
      const url  = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status : 200,
        body   : { foo: 'bar' },
        headers: {
          'Content-Type': 'application/json'
        }
      }, {
        method: 'put'
      });

      return trae.put(url, { foo: 'bar' })
      .then((res) => {
        expect(res).toEqual({ foo: 'bar' });
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('put');
      });
    });
  });

  describe('patch', () => {
    it('makes a PATCH request to baseURL + path', () => {
      const trae = new Trae();
      const url  = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status : 200,
        body   : { foo: 'bar' },
        headers: {
          'Content-Type': 'application/json'
        }
      }, {
        method: 'patch'
      });

      return trae.patch(url, { foo: 'bar' })
      .then((res) => {
        expect(res).toEqual({ foo: 'bar' });
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('patch');
      });
    });
  });

});
