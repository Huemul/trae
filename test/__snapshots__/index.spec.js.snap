exports[`HTTP -> http del makes a DELETE request to baseURL + path 1`] = `
Object {
  "data": "Deleted!",
  "headers": Headers {
    "_headers": Object {
      "content-type": Array [
        "application/text",
      ],
    },
  },
  "status": 200,
}
`;

exports[`HTTP -> http get makes a GET request to baseURL + path 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    "_headers": Object {
      "content-type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
}
`;

exports[`HTTP -> http head makes a HEAD request to baseURL + path 1`] = `
Object {
  "data": "",
  "headers": Headers {
    "_headers": Object {},
  },
  "status": 200,
}
`;

exports[`HTTP -> http patch makes a PATCH request to baseURL + path 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    "_headers": Object {
      "content-type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
}
`;

exports[`HTTP -> http post makes a POST request to baseURL + path 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    "_headers": Object {
      "content-type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
}
`;

exports[`HTTP -> http put makes a PUT request to baseURL + path 1`] = `
Object {
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    "_headers": Object {
      "content-type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
}
`;

exports[`HTTP -> http response not ok gets rejected 1`] = `[Error: Not Found]`;

exports[`trae defaults returns the current default config when no params are passed 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
    "Content-Type": "application/json",
  },
  "xsrfCookieName": "XSRF-TOKEN",
  "xsrfHeaderName": "X-XSRF-TOKEN",
}
`;

exports[`trae defaults sets the default config to be used on all requests for the instance 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
    "Content-Type": "application/json",
  },
  "mode": "no-cors",
  "xsrfCookieName": "XSRF-TOKEN",
  "xsrfHeaderName": "X-XSRF-TOKEN",
}
`;
